web/src/lib/stores/assets-store.svelte.ts

```diff
@@ -162,746 +504,708 @@ export const photoViewerImgElement = writable<HTMLImageElement | null>(null);
+    return this.#findBucketForAsset(assetId);
   }
 
-  async getRandomAsset(): Promise<AssetResponseDto | null> {
-    let index = Math.floor(
-      Math.random() * this.buckets.reduce((accumulator, bucket) => accumulator + bucket.bucketCount, 0),
-    );
-    for (const bucket of this.buckets) {
-      if (index < bucket.bucketCount) {
-        await this.loadBucket(bucket.bucketDate);
-        return bucket.assets[index] || null;
-      }
-
-      index -= bucket.bucketCount;
-    }
+  async getRandomBucket() {
+    const random = Math.floor(Math.random() * this.buckets.length);
+    const bucket = this.buckets[random];
+    await this.loadBucket(bucket.bucketDate, { cancelable: false });
+    return bucket;
+  }
 
-    return null;
+  async getRandomAsset() {
+    const bucket = await this.getRandomBucket();
+    return bucket?.getRandomAsset();
   }
 
-  updateAssets(assets: AssetResponseDto[]) {
-    if (assets.length === 0) {
-      return;
+  // runs op on assets, returns unprocessed
+  #runAssetOp(ids: Set<string>, op: AssetOperation) {
+    if (ids.size === 0) {
+      return { processedIds: new Set(), unprocessedIds: ids, changedGeometry: false };
     }
-    const assetsToRecalculate: AssetResponseDto[] = [];
 
-    for (const _asset of assets) {
-      const asset = this.assets.find((asset) => asset.id === _asset.id);
-      if (!asset) {
-        continue;
-      }
-
-      const recalculate = asset.localDateTime !== _asset.localDateTime;
-      Object.assign(asset, _asset);
-
-      if (recalculate) {
-        assetsToRecalculate.push(asset);
+    const changedBuckets = new Set<AssetBucket>();
+    let idsToProcess = new Set(ids);
+    const idsProcessed = new Set<string>();
+    const combinedMoveAssets: { asset: AssetResponseDto; year: number; month: number }[][] = [];
+    for (const bucket of this.buckets) {
+      if (idsToProcess.size > 0) {
+        const { moveAssets, processedIds, changedGeometry } = bucket.runAssetOp(idsToProcess, op);
+        if (moveAssets.length > 0) {
+          combinedMoveAssets.push(moveAssets);
+        }
+        idsToProcess = idsToProcess.difference(processedIds);
+        for (const id of processedIds) {
+          idsProcessed.add(id);
+        }
+        if (changedGeometry) {
+          changedBuckets.add(bucket);
+          break;
+        }
       }
     }
+    if (combinedMoveAssets.length > 0) {
+      this.#addAssetsToBuckets(combinedMoveAssets.flat().map((a) => a.asset));
+    }
+    const changedGeometry = changedBuckets.size > 0;
+    for (const bucket of changedBuckets) {
+      this.#updateGeometry(bucket, true);
+    }
+    if (changedGeometry) {
+      this.updateIntersections();
+    }
+    return { unprocessedIds: idsToProcess, processedIds: idsProcessed, changedGeometry };
+  }
 
-    this.removeAssets(assetsToRecalculate.map((asset) => asset.id));
-    this.addAssetsToBuckets(assetsToRecalculate);
+  updateAssetOp(ids: string[], op: AssetOperation) {
```

Comment:
Can you help add some comments on when to use this?

```suggestion
  updateAssetOperation(ids: string[], op: AssetOperation) {
```